defuse/src/accounts/account.rs

```diff
@@ -0,0 +1,144 @@
+use defuse_contracts::{
+    crypto::{Payload, PublicKey, SignedPayload},
+    defuse::{payload::ValidatePayloadAs, DefuseError, Result},
+    nep413::Nep413Payload,
+    utils::{cache::CURRENT_ACCOUNT_ID, prefix::NestPrefix, Lock},
+};
+use impl_tools::autoimpl;
+use near_account_id::AccountType;
+use near_sdk::{
+    borsh::BorshSerialize, near, store::IterableMap, AccountId, BorshStorageKey, IntoStorageKey,
+};
+
+use super::{AccountState, Nonces};
+
+#[derive(Debug)]
+#[near(serializers = [borsh])]
+#[autoimpl(Deref using self.state)]
+#[autoimpl(DerefMut using self.state)]
+pub struct Account {
+    /// Nonces used in case of implicit [`AccountId`]
+    implicit_nonces: Lock<Nonces>,
+    public_keys: IterableMap<PublicKey, Lock<Nonces>>,
+
+    pub state: AccountState,
+
+    prefix: Vec<u8>,
+}
+
+impl Account {
+    #[inline]
+    pub fn new<S>(prefix: S, me: &AccountId) -> Self
+    where
+        S: IntoStorageKey,
+    {
+        let prefix = prefix.into_storage_key();
+
+        Self {
+            implicit_nonces: Lock::new(
+                Nonces::new(prefix.as_slice().nest(AccountPrefix::ImplicitNonces)),
+                matches!(me.get_account_type(), AccountType::NamedAccount),
+            ),
+            public_keys: IterableMap::new(prefix.as_slice().nest(AccountPrefix::PublicKeys)),
+            state: AccountState::new(prefix.as_slice().nest(AccountPrefix::State)),
+            prefix,
+        }
+    }
+
+    #[inline]
+    pub fn add_public_key(&mut self, me: &AccountId, public_key: PublicKey) -> &mut Nonces {
```

Comment:
I would simplify the logic of the method by removing the return value. Single responsibility. You have a ` public_key_nonces_mut` for getting value.